src/peft/tuners/lora/layer.py

```diff
@@ -186,8 +191,17 @@ def olora_init(self, adapter_name):
         self.lora_B[adapter_name].weight.data = Qr.contiguous()
 
         weight_tensor.data -= scale_factor * self.lora_B[adapter_name].weight @ self.lora_A[adapter_name].weight
-        weight_tensor = weight_tensor.to(dtype)
-        self.get_base_layer().weight.data = weight_tensor
+        if bnb_param_type == "4bit":
+            weight_tensor = orig_weight.__class__(weight_tensor, quant_type=orig_weight.quant_type).to(
+                orig_weight.device
+            )
+            base_layer.weight = weight_tensor
+        elif bnb_param_type == "8bit":
+            weight_tensor = orig_weight.__class__(weight_tensor, requires_grad=False).to(orig_weight.device)
+            base_layer.weight = weight_tensor
+        else:
+            weight_tensor = weight_tensor.to(dtype)
+            base_layer.weight.data = weight_tensor
```

Comment:
Normally for bnb weights, the tensors are flattened, e.g. shape [64, 1]. But after dequantizing, the `weight_tensor` that we assign here is not flat anymore, e.g. shape [16, 16]. My reasoning was that we should get back a "correct" tensor, so better to reinitialize it.

I tried what happens when I remove this and just do `base_layer.weight.data = weight_tensor` and curiously, this seems to work too and the test passes, even if the shape is now wrong. This makes me wonder if bnb somehow handles this automatically and we should not re-initialize (which could cause its own problems)? Not sure, any suggestion?

Reply:
> I tried what happens when I remove this and just do base_layer.weight.data = weight_tensor

Wow that's really strange indeed. I tried to check the code in bnb and it doesn't look like they handle this. cc @matthewdouglas 

> This makes me wonder if bnb somehow handles this automatically and we should not re-initialize (which could cause its own problems)?

I think that's fine as long as you pass the relevant [kwargs](https://github.com/bitsandbytes-foundation/bitsandbytes/blob/e4674531dd54874c0abbc786ad5635c92c34dc3e/bitsandbytes/nn/modules.py#L423) that you can get from `orig_weight`. However, make sure to not pass `bnb_quantized` arg for the 4-bit case. Then, with to(orig_weight.device), it should quantize the weights properly.

Reply:
Great, thanks for the additional info.

> However, make sure to not pass `bnb_quantized` arg for the 4-bit case. Then, with `to(orig_weight.device)`, it should quantize the weights properly.

To clarify, is the present code in alignment with what you suggest or do I need to call `to(orig_weight.device)` too?

> Wow that's really strange indeed. I tried to check the code in bnb and it doesn't look like they handle this.

Okay, then it's probably better to get Matthew's opinion before merging this.

Reply:
As far as the shapes go, both 4bit and 8bit have some mechanisms in place to track the original shapes, but it's different for each. The `Linear8bitLt` has a `state.SB` and for 4bit that information is part of `quant_state`. The main expectation is that it is all stored in a contiguous row-major format.

That said, it's not really clear to me that `dequantize_module_weight()` is doing all that it would need to do. Maybe it would past the test here but I would think the updated weights would not be quantized properly afterwards, so re-initializing it is probably the way to go.

> To clarify, is the present code in alignment with what you suggest or do I need to call to(orig_weight.device) too?

You'd want to have `.to(orig_weight.device)` in addition to the other kwargs as @SunMarc mentioned.

Reply:
I'm calling `.to(orig_weight.device)` for bnb weights:

- https://github.com/huggingface/peft/pull/2011/files#diff-24a141c266b7b714ae8fcc470f31bc283f7b0f5a671bbf6d5f092741fc374104R196
- https://github.com/huggingface/peft/pull/2011/files#diff-24a141c266b7b714ae8fcc470f31bc283f7b0f5a671bbf6d5f092741fc374104R200

I'm not calling it in the `else` clause, but that's for normal weights. So the code should be fine as is, right?

Reply:
Yes the code is fine as is for the `.to`. What needs to be updated for the kwargs that you are passing to the class. Right now, you are only passing `quant_type=orig_weight.quant_type` for 4-bit and `requires_grad=False` for 8bit.

Reply:
I [updated the inits](https://github.com/huggingface/peft/pull/2011/commits/db62772911237c10e983ccb83868b8bbd5db646f) to take into account all arguments. Unfortunately, this may get out of date if bnb is updated, but I think there is no method such as `bnb.create_param_like(tensor)` or such to offload this work to bnb.

It would be great if you could do a final pass over the change.

Reply:
Can't we just pass `orig_weight.__dict_` as the kwargs ? This is what how we did it in [transformers](https://github.com/huggingface/transformers/blob/28952248b19db29ca25ccf34a5eec413376494a9/src/transformers/quantizers/quantizer_bnb_4bit.py#L230).

Reply:
Hmm, I wonder if that's really more robust. If a new attribute is added that is not an `__init__` argument, this would fail, right?

```python
class Foo:
    def __init__(self, x):
        self.x = x
        self.y = 123

foo = Foo("hi")
Foo(**foo.__dict__)  # TypeError: Foo.__init__() got an unexpected keyword argument 'y'
```

So no matter what, this code may break if there is some change to the `__init__` code in bnb.

Reply:
Oh yeah, that right :/

Reply:
Okay, I merged as is then. Code is going to eventually break one way or the other :D